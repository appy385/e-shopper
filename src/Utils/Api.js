import React from 'react';
import axios from 'axios';

const getCall = async (path) => {
  const res = await axios.get(path);
  return res.data;
};

export default getCall;
