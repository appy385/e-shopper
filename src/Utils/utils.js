import React from 'react';

const groupByCategory = (products, tag) => products.reduce((acc, product) => {
  const newProduct = {
    ...product,
    src: 'https://images.pexels.com/photos/61127/pexels-photo-61127.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=500',
    quantity: (tag === 'products') ? product.count : 0,
    count: (tag === 'products') ? 0 : product.count,

  };
  const { category } = newProduct;
  if (!acc[category]) {
    acc[category] = [];
  }
  acc[category].push(newProduct);
  return acc;
}, {});

export default groupByCategory;
