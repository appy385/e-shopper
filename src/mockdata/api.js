export const mockGetItems = {
  data: {
    data: [
      {
        id: 1,
        name: 'apple',
        price: 120,
        count: 20,
        category: 'Fruits & Vegatables',
      },
      {
        id: 2,
        name: 'table cloth',
        price: 200,
        count: 3,
        category: 'Household Items',
      }],
  },
};

export const mockGetOrders = {
  data: {
    data: [
      {
        items: [
          {
            id: 1,
            name: 'apple',
            price: 120,
            count: 1,
            category: 'Fruits & Vegatables',
          },
        ],
        id: 1,
        date: 1615122360481,
      },
    ],
  },
};

export const mockCreateOrder = {
  data: {
    data: {

      items: [
        {
          id: 4,
          name: 'duster',
          price: 80,
          count: 1,
          category: 'Household Items',
        },
        {
          id: 5,
          name: 'milk',
          price: 10,
          count: 2,
          category: 'Dairy & Eggs',
        },
      ],
    },
  },
};
