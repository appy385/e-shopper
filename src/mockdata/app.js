export const mockAppProducts = {
  'Fruits & Vegatables': [{
    id: 1,
    src: 'https://images.pexels.com/photos/61127/pexels-photo-61127.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=500',
    seller: 'Fresho',
    name: 'Banana-Robusta',
    quantity: 10,
    price: 40,
    count: 0,
    category: 'Fruits & Vegatables',
  },
  {
    id: 2,
    src: 'https://images.pexels.com/photos/61127/pexels-photo-61127.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=500',
    seller: 'Fresho',
    name: 'Banana-Organic',
    quantity: 10,
    price: 40,
    count: 0,
    category: 'Fruits & Vegatables',
  },
  ],
};

export const mockNew = {};
