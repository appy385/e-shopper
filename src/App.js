import React, { useState, useEffect } from 'react';
import './App.scss';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { ThemeContext, Theme } from './ThemeContext';
import { getItems, getOrders } from './Utils/Api';
import groupByCategory from './Utils/Helper';
import HomePage from './Home/HomePage';
import Header from './Header/Header';
import AllOrdersPage from './Order/AllOrdersPage';
import BasketPage from './Basket/BasketPage';
import BasketCheckoutPage from './Basket/BasketCheckoutPage';

const App = () => {
  const [theme, setTheme] = useState('dark');
  const [cartCount, setCartCount] = useState(0);
  const [cartItems, setCartItems] = useState({});
  const [products, setProducts] = useState({});
  const [orders, setOrders] = useState([]);

  useEffect(async () => {
    const productList = await getItems();
    if (productList) {
      const categoryProducts = groupByCategory(productList.data, 'products');
      setProducts(categoryProducts);
    }
  }, []);

  useEffect(async () => {
    const orderList = await getOrders();
    if (orderList) {
      const ordersData = orderList.data;
      const newOrders = ordersData.map((order) => ({
        ...order,
        total: order.items.length,
        cost: order.items.reduce((acc, item) => (acc + item.price * item.count), 0),
        items: groupByCategory(order.items, 'orders'),
      }));
      setOrders(newOrders);
    }
  }, []);

  const onIncrement = (id, category) => {
    setCartCount(cartCount + 1);
    const newCategoryProducts = products[category].map((product) => ((product.id === id) ? {
      ...product,
      count: product.count + 1,
      quantity: product.quantity - 1,
    } : product));

    const newProducts = { ...products };
    newProducts[category] = newCategoryProducts;
    setProducts(newProducts);

    const newCartItems = { ...cartItems };
    newCartItems[category] = newProducts[category].filter((product) => product.count > 0);
    setCartItems(newCartItems);
  };

  const onDecrement = (id, category) => {
    const newCartCount = (cartCount > 0) ? cartCount - 1 : cartCount;
    setCartCount(newCartCount);
    const newCategoryProducts = products[category].map((product) => ((product.id === id) ? {
      ...product,
      count: (product.count > 0) ? product.count - 1 : product.count,
      quantity: (product.count > 0) ? product.quantity + 1 : product.quantity,
    } : product));

    const newProducts = { ...products };
    newProducts[category] = newCategoryProducts;
    setProducts(newProducts);

    const newCartItems = { ...cartItems };
    newCartItems[category] = newProducts[category].filter((product) => product.count > 0);
    setCartItems(newCartItems);
  };

  const handleThemeChange = () => {
    if (theme === 'dark') {
      setTheme('light');
    } else {
      setTheme('dark');
    }
  };

  const addOrder = (order) => {
    const newOrder = {
      ...order,
      total: order.items.length,
      cost: order.items.reduce((acc, item) => (acc + item.price * item.count), 0),
      items: groupByCategory(order.items, 'orders'),
    };
    setOrders([...orders, newOrder]);
  };

  return (
    <BrowserRouter>
      <div>
        <label htmlFor="theme-input">Change Theme mode</label>
        <input id="theme-input" type="checkbox" onClick={handleThemeChange} />

        <ThemeContext.Provider value={(theme === 'dark' ? Theme.dark : Theme.light)}>
          <Header value={cartCount} />
        </ThemeContext.Provider>

        <Switch>
          <Route path="/checkout" exact>
            <BasketCheckoutPage basket={cartItems} addOrder={addOrder} />
          </Route>
          <Route path="/order" exact>
            <AllOrdersPage orders={orders} />
          </Route>
          <Route path="/cart" exact>
            <BasketPage basket={cartItems} />
          </Route>
          <Route path="/">
            <HomePage
              products={products}
              onIncrement={onIncrement}
              onDecrement={onDecrement}
            />
          </Route>
        </Switch>
      </div>
    </BrowserRouter>
  );
};

export default App;
