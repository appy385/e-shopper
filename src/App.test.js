import React from 'react';
import {
  fireEvent, render, screen, act, waitFor,
} from '@testing-library/react';
import { mockGetItems, mockGetOrders } from './mockdata/api';
// import api from './Utils/Api';
import App from './App';
import { getItems, getOrders } from './Utils/Api';

jest.mock('./Utils/Api');

describe(App.name, () => {
  beforeEach(() => {
    getItems.mockResolvedValue(mockGetItems.data);
    getOrders.mockResolvedValue(mockGetOrders.data);
  });
  afterEach(() => {
    jest.clearAllMocks();
    getOrders.mockClear();
    getItems.mockClear();
  });
  test('should match snapshot', async () => {
    const { container } = await waitFor(() => render(<App />));
    expect(container).toMatchSnapshot();
  });

  test('should display input checkbox field', async () => {
    await waitFor(() => render(<App />));
    const themeInput = screen.getByLabelText('Change Theme mode');
    expect(themeInput.tagName).toBe('INPUT');
  });

  test('should turn input checkbox field to true on click', async () => {
    await waitFor(() => render(<App />));
    const themeCheckbox = screen.getByLabelText('Change Theme mode');
    fireEvent.click(themeCheckbox);
    expect(themeCheckbox.checked).toEqual(true);
  });

  test('should render order page', async () => {
    await waitFor(() => render(<App />));
    const orderElement = screen.getByText('All Orders');
    act(() => { fireEvent.click(orderElement); });
    expect(document.location.href).toBe('http://localhost/order');
  });

  test('should render cart page', async () => {
    await waitFor(() => render(<App />));
    const cartElement = screen.getByText('Basket Items:');
    act(() => { fireEvent.click(cartElement); });
    expect(document.location.href).toBe('http://localhost/cart');
  });

  test('should increase count by 1 when + button is clicked', async () => {
    await waitFor(() => render(<App />));
    const homeElement = screen.getByText('E-shopper');
    act(() => { fireEvent.click(homeElement); });
    expect(document.location.href).toBe('http://localhost/');
    const incrementButton = screen.getAllByText('+');
    act(() => { fireEvent.click(incrementButton[0]); });
    screen.getByText('1');
  });

  it('should get orders data from backend', async () => {
    await waitFor(() => render(<App />));
    expect(getOrders).toHaveBeenCalledTimes(1);
  });

  it('should get items data from backend', async () => {
    await waitFor(() => render(<App />));
    expect(getItems).toHaveBeenCalledTimes(1);
  });
});
