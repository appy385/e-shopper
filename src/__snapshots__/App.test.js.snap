// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <div>
    <label
      for="theme-input"
    >
      Change Theme mode
    </label>
    <input
      id="theme-input"
      type="checkbox"
    />
    <div
      class="header-dark"
    >
      <a
        href="/"
      >
        <p
          class="header-tab-dark eshopper"
        >
          E-shopper
        </p>
      </a>
      <div
        class="header-right"
      >
        <a
          href="/order"
        >
          <p
            class="header-tab-dark orders"
          >
            All Orders
          </p>
        </a>
        <a
          href="/cart"
        >
          <div
            class="header-basket-container"
          >
            <img
              alt="logo"
              class="header-basket-image"
              src="basket.png"
            />
            <p
              class="header-tab-dark items"
            >
              Basket Items:
            </p>
            <p
              class="header-tab-dark items"
            >
              0
            </p>
          </div>
        </a>
      </div>
    </div>
    <div
      class="home"
    >
      <div
        class="home-page"
      >
        <div
          class="home-category"
        >
          <strong>
            Fruits & Vegatables
          </strong>
          <div
            class="home-category-row"
            data-testid="home-category-row"
          >
            <div
              class="product-card"
              data-testid="product-card"
            >
              <img
                alt="product"
                src="https://images.pexels.com/photos/61127/pexels-photo-61127.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=500"
              />
              <div
                class="product-seller"
              />
              <div
                class="product-name"
              >
                apple
              </div>
              <div
                class="product-quantity"
              >
                In stock: 20
              </div>
              <div
                class="product-card-box"
              >
                <div
                  class="product-price"
                >
                  MRP: 120 /-
                </div>
                <div
                  class="quantity-container"
                >
                  <button
                    class="quantity-inc-button"
                    type="button"
                  >
                    +
                  </button>
                  <span
                    class="quantity-basket-count"
                  >
                    0 in Basket
                  </span>
                  <button
                    class="quantity-dec-button"
                    type="button"
                  >
                    -
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="home-category"
        >
          <strong>
            Household Items
          </strong>
          <div
            class="home-category-row"
            data-testid="home-category-row"
          >
            <div
              class="product-card"
              data-testid="product-card"
            >
              <img
                alt="product"
                src="https://images.pexels.com/photos/61127/pexels-photo-61127.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=500"
              />
              <div
                class="product-seller"
              />
              <div
                class="product-name"
              >
                table cloth
              </div>
              <div
                class="product-quantity"
              >
                In stock: 3
              </div>
              <div
                class="product-card-box"
              >
                <div
                  class="product-price"
                >
                  MRP: 200 /-
                </div>
                <div
                  class="quantity-container"
                >
                  <button
                    class="quantity-inc-button"
                    type="button"
                  >
                    +
                  </button>
                  <span
                    class="quantity-basket-count"
                  >
                    0 in Basket
                  </span>
                  <button
                    class="quantity-dec-button"
                    type="button"
                  >
                    -
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
